File: apps/scala/src/main/scala/concurrency/ch13/WordCount.scala
package concurrency.ch13

import scala.collection.parallel.CollectionConverters.*

object WordCount:

  /** Map: Convert text to (word, 1) pairs */
  def map(text: String): List[(String, Int)] =
    text.toLowerCase
      .split("\\s+")
      .filter(_.nonEmpty)
      .map(word => (word, 1))
      .toList

  /** Reduce: Aggregate word counts */
  def reduce(pairs: List[(String, Int)]): Map[String, Int] =
    pairs.groupBy(_._1).view.mapValues(_.map(_._2).sum).toMap

  /** MapReduce: Count words in multiple texts using parallel collections */
  def countWords(texts: List[String]): Map[String, Int] =
    // Map phase (parallel)
    val mapped = texts.par.flatMap(map).toList

    // Reduce phase
    reduce(mapped)

File: apps/haskell/src/Ch13/WordCount.hs
module Ch13.WordCount
    ( mapPhase
    , reducePhase
    , countWords
    , countWordsParallel
    ) where

import Control.Concurrent.Async (mapConcurrently)
import Data.Char (toLower, isAlphaNum)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.List (foldl')

-- | Map: Convert text to (word, 1) pairs
mapPhase :: String -> [(String, Int)]
mapPhase text =
    [(map toLower word, 1) | word <- words text, not (null word)]

-- | Reduce: Aggregate word counts
reducePhase :: [(String, Int)] -> Map String Int
reducePhase = foldl' addPair Map.empty
  where
    addPair acc (word, count) = Map.insertWith (+) word count acc

-- | MapReduce: Count words in multiple texts (sequential)
countWords :: [String] -> Map String Int
countWords texts =
    let mapped = concatMap mapPhase texts
    in reducePhase mapped

-- | MapReduce: Count words using parallel execution
countWordsParallel :: [String] -> IO (Map String Int)
countWordsParallel [] = return Map.empty
countWordsParallel texts = do
    -- Map phase (parallel)
    mapped <- mapConcurrently (return . mapPhase) texts
    -- Reduce phase
    return $ reducePhase (concat mapped)
